### DISCUSSION TITLE
[Feature Request] Port agent to Golang or Rust

### DISCUSSION THREAD

**Discussion by Hybrid512** [glpi-project/glpi-agent#363] [Ideas]
GLPI is great but the agent is really problematic due to its dependencies to Perl version and perl modules that are not always available on every distribution.
A Golang or Rust based version would be really great.
Basically, something statically built which don't need any extra dependency and that can be cross compiled easily the same way this is done for most Prometheus exporters where deploying such agent is just a matter of deploying the binary somewhere in the PATH and have a systemD unit or init script to start it as a daemon.
It would also work way better with non Linux builds such as Windows or Mac OS X, BSDs, ... and even non x86 arch.

**Comment by g-bougard** [comment]
Hi @Hybrid512 

totally agree with you, and you should know this will happen. [As still announced](https://glpi-project.org/glpi-news-and-work-in-progress-fall-2022/), we still choose to use Golang to port the agent:

> **GLPI Agent – Roadmap (2/2)** 
> 
> We thought about rewriting a large part of the agents, and we shortlisted Golang for that. 
> 
> The fact the agent is written with the language PERL. 
> 
> Recently, finding developers comfortable with this language has been hard. 
> 
> We will try a prototype within the year to measure our capacity to switch to this new language. 
> 
> We plan also to enforce exchange between agents and backend by adding authentication and registration flows. This will be recommended but you will be able to do simple exchanges if you prefer. 
> 
> And we still need to redo all forms to drive remotely the agent in GLPI UI. 
> 
> This is still planned. 

Write all the agent in Go with all the supported tasks will take a very long time. We will probably rewrite the agent core in a first stage, giving it the ability to support other languages, probably python, maybe javascript too. This may help many people to provide new checks. This should be GLPI-Agent 2.0.
Having everything written in Go could a second stage and this would be GLPI-Agent 3.0.

**Comment by tetsumaki** [comment]
zabbix-agent2 has also been rewritten in go to facilitate deployment, development and more.  
python also causes problems (dependancies/deployment/consumption) with versions 3.8, 3.9, 3.10, 3.11, 3.12 related to os (rhel, debian, ...)
glpi-agent written in go or rust is a very good idea.
